File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

/* class Solution {
    public boolean hasCycle(ListNode head) {
        while (head != null) {
            ListNode current = head.next;
            List<ListNode> visited = new ArrayList<>();
            while (current != null && !visited.contains(current)) {
                if(current == head) {
                    return true;
                }
                visited.add (current);
                current = current.next;
            }
            head = head.next;
        }
        return false;
    }
} */

class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                return true;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
          # part 1: take 2nd portion of list and reverse it
          # part 2: join the lists

          # find middle of list by using fast and slow pointers
          # when fast reaches the end, slow will be in the middle
          # need 2 temp pointers so we dont break the link

          slow = head
          fast = head.next 
          while fast and fast.next: # while fast is not at end of list
            slow = slow.next
            fast = fast.next.next

          second = slow.next # second half of the list
          prev = slow.next = None

          while second: #reversing second portion of list
            temp = second.next
            second.next = prev
            prev = second
            second = temp

        # merge two halfs
          first = head
          second = prev
          while second: # because second half is shorter or equal to first
            temp1 = first.next
            temp2 = second.next
            first.next = second
            second.next = temp1
            # reset pointers
            first = temp1
            second = temp2
        # dont have to return because we changed in-place

        